File: mindy/week4/O-4949.py
# 문자의 균형 맞추기
# 괄호 맞추기
# 이전값 다음값을 저장해서 



    
while(True):
    stack =[]
    
    s = input()
    if s =='.':
        break
    
    
    for i in s:
        if i == '(' or i=='[':
            stack.append(i)
        elif i == ')':
            if stack and stack[-1]=='(':
                stack.pop()
            else:
                stack.append(i)
                break
        elif i == ']':
            if stack and stack[-1]=='[':
                stack.pop()
            else:
                stack.append(i)
                break
    
    if not stack:
            print("yes")
    else:
            print("no")
            
        

        

File: leo/week5/O-4949.js
/**
 *  제목 :
 *  입력 :
 *  제한사항 :
 *  시간계산 필요 :
 */

// Run by Node.js
const readline = require("readline");

(async () => {
  let rl = readline.createInterface({ input: process.stdin });

  let input = [];
  for await (const line of rl) {
    input.push(line);
  }
  rl.close();

  input.forEach((line) => {
    let stack = [];
    let flag = true;
    let count = 0;
    for (ch of line) {
      switch (ch) {
        case "(":
          stack.push("(");
          ++count;
          break;
        case ")":
          if (stack.at(-1) === "(") stack.pop();
          else flag = false;
          break;
        case "[":
          stack.push("[");
          ++count;
          break;
        case "]":
          if (stack.at(-1) === "[") stack.pop();
          else flag = false;
          break;
        case ".":
          if (count) {
            console.log(stack.length === 0 ? "yes" : "no");
          }
          break;
        default:
          ++count;
      }
      if (!flag) {
        console.log("no");
        break;
      }
    }
  });

  process.exit();
})();
